Allow config :log to be explicitly set to `stdout`

tombruijn opened the pull request:
Will silence any warnings of not writable log file paths when set to
`log: stdout`.

Part of #184
Doesn't fix it entirely as it only redirects the ruby gem's output.

This allows the user to have some choice in the matter where
things are logged.
It not perfect system, but expect a RFC for the agent's side.

tombruijn commented on lib/appsignal.rb:
```diff
@@ -301,6 +294,15 @@ def start_stdout_logger
         "appsignal: #{msg}\n"
```
@thijsc I kept this intact, but wouldn't it be more useful to use the normal `log_formatter` here too?

thijsc replied:
This is because Heroku already appends timestamps to sdout logs. Not sure how this works on other platforms.

thijsc replied:
Not sure how this is usually handled. Maybe it's normal that timestamps are already appended on most platforms?

tombruijn replied:
I think it's quite normal an application would log its own logs with a timestamp. That Heroku adds a timestamp to all logs is something that's part of their logging system and only because it's wrapped in a foreman-like.

I'm 👍  for adding a timestamp to it, not making an exception for Heroku to stay consistent.

thijsc replied:
ok

tombruijn commented:
Pushed a change that changes the behavior slightly.

Previously (in this PR) `Appsignal::Config#log_file_path` would return `nil` when `log: stdout` was set, but this would cause the Rust agent to not know where to write its log file to and revert to using the `/tmp/appsignal.log` path.

Now it `Appsignal::Config#log_file_path` returns a value again, regardless of the `log` option. It's only used by the Rust agent, with `log: stdout`, until we have support for amending its log to the Ruby process'. This means that it will still output warnings when the log path is not writable even if `log` is set to `stdout`, because we can't fully support it yet.

tombruijn commented:
Build on the agent side is working, so the logic is sane there too.

Please re-review @thijsc @jeffkreeftmeijer @matsimitsu

matsimitsu commented:
> Now it Appsignal::Config#log_file_path returns a value again, regardless of the log option. It's only used by the Rust agent, with log: stdout, until we have support for amending its log to the Ruby process'. This means that it will still output warnings when the log path is not writable even if log is set to stdout, because we can't fully support it yet.

Does the agent continue booting when the path is not writable?

tombruijn commented:
Looking at [agent main.rs](https://github.com/appsignal/appsignal-agent/blob/2ab3d1464e1bf0f86303b62387db3454ec36201e/agent/src/main.rs#L51-L61) I'm going to say no. It exits when there is no writable log path, not the path configured by the user or the fallback `/tmp/appsignal.log`. This behavior will remain unchanged.

tombruijn commented:
@thijsc I still had a `fixup` commit (https://robots.thoughtbot.com/autosquashing-git-commits) in the commits, meant to be squashed with the other commit in a rebase. I used a fixup commit instead of amending my changes to not break the diff of new changes since the last commit on GitHub.

Next time I'll put a warning sign up when the PR needs to be merged with a rebase.
